nbconvert/preprocessors/execute.py

```diff
@@ -466,55 +466,59 @@ def run_cell(self, cell, cell_index=0):
 
-            msg_type = msg['msg_type']
-            self.log.debug("output: %s", msg_type)
-            content = msg['content']
-
-            # set the prompt number for the input and the output
-            if 'execution_count' in content:
-                cell['execution_count'] = content['execution_count']
-
-            if msg_type == 'status':
-                if content['execution_state'] == 'idle':
-                    break
-                else:
-                    continue
-            elif msg_type == 'execute_input':
-                continue
-            elif msg_type == 'clear_output':
-                outs[:] = []
-                # clear display_id mapping for this cell
-                for display_id, cell_map in self._display_id_map.items():
-                    if cell_index in cell_map:
-                        cell_map[cell_index] = []
-                continue
-            elif msg_type.startswith('comm'):
-                continue
-
-            display_id = None
-            if msg_type in {'execute_result', 'display_data', 'update_display_data'}:
-                display_id = msg['content'].get('transient', {}).get('display_id', None)
-                if display_id:
-                    self._update_display_id(display_id, msg)
-                if msg_type == 'update_display_data':
-                    # update_display_data doesn't get recorded
-                    continue
-
+            if not self.process_message(msg, cell, cell_index):
+                break
+
+        return exec_reply, cell.outputs
+
+    def process_message(self, msg, cell, cell_index):
+        '''
+        Returns None if execution should be halted.
+        '''
+        msg_type = msg['msg_type']
+        self.log.debug("msg_type: %s", msg_type)
+        content = msg['content']
+        self.log.debug("content: %s", content)
+
+        # Default to our input as the "result" of processing the message
+        result = msg
+
+        display_id = content.get('transient', {}).get('display_id', None)
+        if display_id and msg_type in {'execute_result', 'display_data', 'update_display_data'}:
+            self._update_display_id(display_id, msg)
+
+        # set the prompt number for the input and the output
+        if 'execution_count' in content:
+            cell['execution_count'] = content['execution_count']
+
+        if msg_type == 'status':
+            if content['execution_state'] == 'idle':
+                # Set result to None to halt execution
+                result = None
+        elif msg_type == 'clear_output':
+            cell.outputs[:] = []
+            # clear display_id mapping for this cell
+            for display_id, cell_map in self._display_id_map.items():
+                if cell_index in cell_map:
+                    cell_map[cell_index] = []
+        # Check for remaining messages we don't process
+        elif not (msg_type in ['execute_input', 'update_display_data'] or msg_type.startswith('comm')):
```

Comment:
It's actually identical logically, by not returning it will continue here

Reply:
Might have missed the not?

Reply:
Additionally now if something is an `update_display_data` message, it will return something (specifically the content of the msg). 

I'm starting to wonder if this all might be fixed if we default to returning `None`…

Reply:
Was working fine post-rebase, but I changed this to a raise control flow pattern which I think is a lot easier to read.